### ISSUE TITLE
Manual NWB file validation fails

### ISSUE THREAD

**Issue by t-b** [NeurodataWithoutBorders/pynwb#1082]
## Steps to Reproduce

hdmf: 9a5e2e71 (add append and extend for Datasets in Data objects (#161), 2019-10-02)
pynwb: 5c6c2028 (Added new case to the iterative write tutorial (#1029), 2019-10-02)

```
tox -e gallery

$ python -m pynwb.validate example_file_path.nwb
Traceback (most recent call last):
  File "C:\users\thomas\Anaconda3\lib\runpy.py", line 183, in _run_module_as_main
    mod_name, mod_spec, code = _get_module_details(mod_name, _Error)
  File "C:\users\thomas\Anaconda3\lib\runpy.py", line 109, in _get_module_details
    __import__(pkg_name)
  File "e:\projekte\pynwb\src\pynwb\__init__.py", line 15, in <module>
    import hdmf.common
  File "e:\projekte\hdmf\src\hdmf\common\__init__.py", line 108, in <module>
    raise RuntimeError("Unable to load a TypeMap - no namespace file found")
RuntimeError: Unable to load a TypeMap - no namespace file found
(pynwb)
```

I thought to have prevented errors like this, by having added

https://github.com/NeurodataWithoutBorders/pynwb/blob/5c6c20283bdd5beb6582007dac9c5deca41b580e/test.py#L77-L107

but that seems to be working differently.

## Environment
$ python --version
Python 3.7.3
Windows, Conda

**Comment by bendichter** [comment]
@t-b I think this is a git submodule issue. Would you mind running:

```bash
$ git clone --recurse-submodules git@github.com:NeurodataWithoutBorders/pynwb.git
```
and 
```bash
$ git clone --recurse-submodules git@github.com:hdmf-dev/hdmf.git
```

**Comment by t-b** [comment]
@bendichter 

Thanks,
yes some of them were missing.

I did

```
git submodule init
git submodule update --checkout --force 
```

and now get a different error message

```
$ python -m pynwb.validate example_file_path.nwb
Validating example_file_path.nwb against cached namespace information using namespace hdmf-common.
Traceback (most recent call last):
  File "C:\users\thomas\Anaconda3\lib\runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "C:\users\thomas\Anaconda3\lib\runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "e:\projekte\pynwb\src\pynwb\validate.py", line 117, in <module>
    main()
  File "e:\projekte\pynwb\src\pynwb\validate.py", line 111, in main
    ret = ret or _validate_helper(io=io, namespace=ns)
  File "e:\projekte\pynwb\src\pynwb\validate.py", line 26, in _validate_helper
    errors = validate(**kwargs)
  File "e:\projekte\hdmf\src\hdmf\utils.py", line 453, in func_call
    return func(**parsed['args'])
  File "e:\projekte\pynwb\src\pynwb\__init__.py", line 179, in validate
    return validator.validate(builder)
  File "e:\projekte\hdmf\src\hdmf\utils.py", line 438, in func_call
    return func(self, **parsed['args'])
  File "e:\projekte\hdmf\src\hdmf\validate\validator.py", line 237, in validate
    validator = self.get_validator(dt)
  File "e:\projekte\hdmf\src\hdmf\utils.py", line 438, in func_call
    return func(self, **parsed['args'])
  File "e:\projekte\hdmf\src\hdmf\validate\validator.py", line 222, in get_validator
    raise_from(ValueError(msg), None)
  File "<string>", line 3, in raise_from
ValueError: data type 'NWBFile' not found in namespace hdmf-common
```

**Comment by bendichter** [comment]
@t-b you need to run it for hdmf and for pynwb

**Comment by t-b** [comment]
I did that of course.

**Comment by t-b** [comment]
Does the file [example_file_path.zip](https://github.com/NeurodataWithoutBorders/pynwb/files/3691053/example_file_path.zip) validate on your end?

**Comment by bendichter** [comment]
```python
from pynwb import NWBHDF5IO, validate

fpath = '/Users/bendichter/Downloads/example_file_path.nwb'

io = NWBHDF5IO(fpath,'r')
validate(io)
```

`[]`

from `pip list`:
```
pynwb                   1.1.0.post0.dev2            /Users/bendichter/dev/pynwb/src                  
hdmf                    1.3.2.post0.dev8            /Users/bendichter/dev/hdmf/src                   
```

**Comment by t-b** [comment]
@bendichter If I do it with you example it works but using it just from the CLI does fail:

```
(base) E:\projekte\pynwb>python -m pynwb.validate example_file_path.nwb
Validating example_file_path.nwb against cached namespace information using namespace hdmf-common.
Traceback (most recent call last):
  File "C:\Users\thomas\Anaconda3\lib\runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "C:\Users\thomas\Anaconda3\lib\runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "e:\projekte\pynwb\src\pynwb\validate.py", line 117, in <module>
    main()
  File "e:\projekte\pynwb\src\pynwb\validate.py", line 111, in main
    ret = ret or _validate_helper(io=io, namespace=ns)
  File "e:\projekte\pynwb\src\pynwb\validate.py", line 26, in _validate_helper
    errors = validate(**kwargs)
  File "C:\Users\thomas\Anaconda3\lib\site-packages\hdmf\utils.py", line 453, in func_call
    return func(**parsed['args'])
  File "e:\projekte\pynwb\src\pynwb\__init__.py", line 179, in validate
    return validator.validate(builder)
  File "C:\Users\thomas\Anaconda3\lib\site-packages\hdmf\utils.py", line 438, in func_call
    return func(self, **parsed['args'])
  File "C:\Users\thomas\Anaconda3\lib\site-packages\hdmf\validate\validator.py", line 237, in validate
    validator = self.get_validator(dt)
  File "C:\Users\thomas\Anaconda3\lib\site-packages\hdmf\utils.py", line 438, in func_call
    return func(self, **parsed['args'])
  File "C:\Users\thomas\Anaconda3\lib\site-packages\hdmf\validate\validator.py", line 222, in get_validator
    raise_from(ValueError(msg), None)
  File "<string>", line 3, in raise_from
ValueError: data type 'NWBFile' not found in namespace hdmf-common

(base) E:\projekte\pynwb>python t.py
[]

(base) E:\projekte\pynwb>
```

```
from pynwb import NWBHDF5IO, validate

fpath = 'example_file_path.nwb'

io = NWBHDF5IO(fpath,'r')
print(validate(io))
```

**Comment by bendichter** [comment]
@t-b that's strange, I would have thought they would act the same. Could it be that your paths are different in these two cases?

**Comment by t-b** [comment]
And I get the exact same error on linux as well.

**Comment by bendichter** [comment]
ok thanks for checking that, looks like this needs some investigation

**Comment by rly** [comment]
The difference is that in `validate.main`, the file is validated against _all_ namespaces cached within the file, whereas calling just the `validate` function imported from pynwb validates the file against only the core namespace.

https://github.com/NeurodataWithoutBorders/pynwb/blob/c6ec626117af813449342ac0954e70d8117829b1/src/pynwb/validate.py#L108-L111

I don't know why that throws an error. 

That's one issue. Next, if I use `validate.main` and specify that I want to validate against only the core namespace, I get a separate issue:
```
AttributeError: 'NWBGroupSpec' object has no attribute 'shape'
```
I suspect this has to do with the fact that a BuildManager is passed into the IO object for `validate.main`.

If I change the manager to be just the default `get_manager` from `pynwb.__init__`, then the file validates against core using `validate.main`. However, it still does not validate against hdmf-common:
```
ValueError: builder must have data type defined with attribute 'data_type'
```
The builder in question is the builder for the root node, an NWBFile. It does not have a `'data_type'` but it does have a `'neurodata_type'`. It seems the wrong type key is used?

@ajtritt since you might be able to follow the logic of the validator better than me, could you look into this?

**Comment by t-b** [comment]
Fixed by #1094. Please reopen if you disagree.
